koku/api/report/ocp_aws/ocp_aws_query_handler.py

```diff
@@ -0,0 +1,123 @@
+#
+# Copyright 2018 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+"""OCP Query Handling for Reports."""
+import copy
+
+from django.db.models import F, Sum, Window
+from django.db.models.functions import (Coalesce, RowNumber)
+from tenant_schemas.utils import tenant_context
+
+from api.report.aws.aws_query_handler import AWSReportQueryHandler
+
+
+class OCPAWSReportQueryHandler(AWSReportQueryHandler):
+    """Handles report queries and responses for OCP on AWS."""
+
+    def __init__(self, query_parameters, url_data,
+                 tenant, **kwargs):
+        """Establish OCP report query handler.
+
+        Args:
+            query_parameters    (Dict): parameters for query
+            url_data        (String): URL string to provide order information
+            tenant    (String): the tenant to use to access CUR data
+            kwargs    (Dict): A dictionary for internal query alteration based on path
+        """
+        kwargs['provider'] = 'OCP_AWS'
+        super().__init__(query_parameters, url_data,
+                         tenant, **kwargs)
+        self.update_cost_field()
+
+    def update_cost_field(self):
+        """Update which field is used to calculate cost by group by param."""
+        group_by = self._get_group_by()
+        self._mapper = copy.deepcopy(self._mapper)
+        if group_by and group_by[0] == 'project':
+            self._mapper._report_type_map['aggregates']['cost'] = Sum('pod_cost')
+            self._mapper._report_type_map['annotations']['cost'] = Sum('pod_cost')
```

Comment:
You shouldn't be updating the map in the code. If the map isn't right, fix the map.

The whole point of having it is to separate the data fields from the code. So, it should be considered read-only.

Reply:
project cost is a special case where we only want to the sum of the pod specific costs to come through. Other times we want to just sum the AWS costs (for node, or cluster, or account, et cetera...) So we'll have to at least update what part of the map we're pulling from dynamically based on the group by argument of this API.

Reply:
Yep. That can be two keys in a dict where you conditionally select which key to use.

It's just the reverse of what you have. :slightly_smiling_face: